### PR TITLE
Refactor `Linearizer`, convert methods to async and use an async context manager

### PR DESCRIPTION
**Pull request by squahtx**
Makes #12114 easier to fix.

Best reviewed commit by commit.

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

File: synapse/util/async_helpers.py [hunk 1]
```diff
@@ -428,56 +414,72 @@ def _ctx_manager() -> Iterator[None]:
...
+            # ensure that we fall back to the reactor between each iteration.
+            #
+            # (This needs to happen while we hold the lock, and the context manager's
+            # exit code must be synchronous, so this is the only sensible place.)
+            await self._clock.sleep(0)
+        else:
+            logger.debug(
+                "Acquired uncontended linearizer lock %r for key %r", self.name, key
+            )
+            entry.count += 1
```
Comment (richvdh) [inline synapse/util/async_helpers.py:459]: it'd be nice to do an early return here, rather than have a big `if` block which is hard to follow.

**Review by richvdh — APPROVED** [review]
lgtm otherwise!

File: synapse/util/async_helpers.py [hunk 2]
```diff
@@ -461,30 +449,35 @@ def cb(_r: None) -> "defer.Deferred[None]":
             # In order to break the cycle, we add a cheeky sleep(0) here to
             # ensure that we fall back to the reactor between each iteration.
             #
-            # (This needs to happen while we hold the lock, and the context manager's exit
-            # code must be synchronous, so this is the only sensible place.)
-            return self._clock.sleep(0)
+            # (This needs to happen while we hold the lock, and the context manager's
+            # exit code must be synchronous, so this is the only sensible place.)
```
Comment (richvdh) [inline synapse/util/async_helpers.py:453]: it's worth noting this is no longer strictly true - the context manager's `__aexit__` doesnt have to be synchronous, so the sleep could happen there instead.

Not sure it's worth changing, just noting that the comment is incorrect.

Comment (squahtx) [reply synapse/util/async_helpers.py]: Good spot! I'll update the comment.

**Comment by squahtx** [comment]
@richvdh Let me know if these changes are what you had in mind!

File: synapse/util/async_helpers.py [hunk 3]
```diff
@@ -402,61 +402,58 @@ async def _acquire_lock(self, key: Hashable) -> _LinearizerEntry:
...
+            logger.info("defer %r got err %r", new_defer, e)
+            if isinstance(e, CancelledError):
+                logger.debug(
+                    "Cancelling wait for linearizer lock %r for key %r",
+                    self.name,
+                    key,
+                )
+            else:
+                logger.warning(
+                    "Unexpected exception waiting for linearizer lock %r for key " "%r",
```
Comment (richvdh) [inline synapse/util/async_helpers.py:434]:
```suggestion
                    "Unexpected exception waiting for linearizer lock %r for key %r",
```

**Review by richvdh — APPROVED** [review]
lgtm otherwise

File: synapse/util/async_helpers.py (hunk 3 above)
Comment (squahtx) [reply synapse/util/async_helpers.py]: black strikes again!
